### PR TITLE
AST: fix interpolated regex tokens

### PR DESCRIPTION
**Pull request by helixbass**
@GeoffreyBooth I turned on the ESLint plugin against the Coffeescript codebase (on a branch) and am using that to look for issues

Here's one on the Coffeescript side:

The tokens for interpolated regexes needed to be cleaned up, this was causing rules like [`keyword-spacing`](https://eslint.org/docs/rules/keyword-spacing) to break

Based on `ast-clean-up-tests`, [here](https://github.com/helixbass/copheescript/compare/ast-clean-up-tests...ast-fix-interpolated-regex-tokens) is just the diff against that branch

### CODE REVIEW

File: src/lexer.coffee [hunk 1]
```diff
@@ -480,16 +480,16 @@ exports.Lexer = class Lexer
         @validateUnicodeCodePointEscapes body, {delimiter}
         @token 'REGEX', "/#{body}/#{flags}", {length: end, origin, data: {delimiter}}
       else
-        @token 'REGEX_START', '(',    {length: 0, origin}
-        @token 'IDENTIFIER', 'RegExp', length: 0
-        @token 'CALL_START', '(',      length: 0
+        @token 'REGEX_START', '(',    {length: 0, origin, generated: yes}
```
Comment (helixbass) [inline src/lexer.coffee]: Need to mark all these fake tokens from converting an interpolated regex into a nonexistent `RegExp()` call as generated

File: src/lexer.coffee [hunk 2]
```diff
@@ -480,16 +480,16 @@ exports.Lexer = class Lexer
...
+        @token 'CALL_START', '(',      length: 0, generated: yes
         @mergeInterpolationTokens tokens, {double: yes, heregex: {flags}, endOffset: end - flags.length, quote: '///'}, (str) =>
           @validateUnicodeCodePointEscapes str, {delimiter}
         if flags
-          @token ',', ',',                    offset: index - 1, length: 0
-          @token 'STRING', '"' + flags + '"', offset: index - 1, length: flags.length
-        @token ')', ')',                      offset: end,       length: 0
-        @token 'REGEX_END', ')',              offset: end,       length: 0
+          @token ',', ',',                    offset: index - 1, length: 0, generated: yes
+          @token 'STRING', '"' + flags + '"', offset: index,     length: flags.length
```
Comment (helixbass) [inline src/lexer.coffee]: The location data generated for interpolated regex flags was off by one

File: src/lexer.coffee [hunk 3]
```diff
@@ -1008,7 +1008,7 @@ exports.Lexer = class Lexer
           lastToken[2].range[1]
         ]
       ]
-      lparen[2] = lparen.origin[2]
+      lparen[2] = lparen.origin[2] unless quote?.length
```
Comment (helixbass) [inline src/lexer.coffee]: I traced this back [here](https://github.com/jashkenas/coffeescript/pull/4572/files#r130409263) to your comments PR

There I think you were running into `lparen` sometimes not having *any* location data set

But it generally does, and this line was wiping out the accurate location data (eg for the opening `///` in an interpolated regex) and replacing it with the location data for the entire interpolated string/regex

So updated it to only "overwrite" `lparen`'s location data if it doesn't already have legitimate location data

File: test/location.coffee [hunk 4]
```diff
@@ -607,9 +607,9 @@ test "Verify heregexes with interpolations have the right ending position", ->
   arrayEndColumn = arrayEnd[2].first_column
   for token in [comma]
     assertColumn token, arrayEndColumn - 2
-  for token in [flagsString]
-    assertColumn token, arrayEndColumn - 2, 1
-  for token in [regexCallEnd, regexEnd, fnCallEnd]
+  for token in [flagsString, regexCallEnd]
```
Comment (helixbass) [inline test/location.coffee]: By updating the `regexCallEnd` token to be marked (correctly) as `generated`, its location data now gets updated by the rewriter (along with other generated tokens)

Its location data isn't particularly significant anyway as it's not a real token

File: src/lexer.coffee (hunk 3 above)
Comment (GeoffreyBooth) [reply src/lexer.coffee]: Always nice to have my own bugs pointed out and fixed 😄

**Comment by GeoffreyBooth** [comment]
✅ once conflicts resolved.

**Comment by helixbass** [comment]
@GeoffreyBooth ok rebased

**Pull request merged**